### DISCUSSION TITLE
Low power modes

### DISCUSSION THREAD

**Discussion by matou78** [SpenceKonde/DxCore#308] [Q&A] [answered]
Hello everyone ! Need some help if possible, with the low power modes .

I'm playing with some avr64DA32 + NRF24L01 to make myself a little remote control with dimming for lights and stuff.

I have already all the transceiver and receiver working as i need but i have trouble to reduce the power consumption.

There is what i did to get the to 0.8mA ( but it's not low enough )
- Internal clock is at 1Mhz
- NRF24L01 is powered only when something needs to be sent
- Hardware reduced to minimum : buttons (using input_pullup) , 1 Led (blink on button action) and the NRF24L01
- all unused pins , set them up to output

The resistance to see the power use is 5.6 Ohm
So around 3.2mA (18.80mV) on action, 1.8mA (10.65mV) on idle and 0.8mA (5.56mV) on sleep .

To put the AVR64DA to sleep , i used just :
-In setup
set_sleep_mode(SLEEP_MODE_PWR_DOWN);
sleep_enable();

- in loop 
sleep_cpu();

I saw that in the datasheet, i should be able to go as low as 0.9 µA in Power-Down mode.
I don't think that i will be able to go that low as DxCore and other things are not compatible with the 32Khz clock.

I saw in the datasheet "VREGCTRL.PMODE" and "VREGCTRL.HTLLEN" that maybe can do few things ? but no idea how to use them .

Any one have ideas or have ben able to do the very low power usage ? or there is no solutions for this as today ?

Thank you for your brain time :)
Have a nice day

**Comment by ObviousInRetrospect** [comment]
Is your measurement just the mcu or the whole board? Is there a voltage regulator? can you share a schematic (if full board). I would suggest measuring just the MCU if you can.

I haven't measured power-down that carefully but I have a board with a 64DB32 that when waking up at 8hz and reading an ina3221 (which itself draws ~375UA) runs around 800ua average across standby and active so 800ua in pure sleep sounds way off. Any chance you have the ADC running?

FWIW you might want to play with measuring power use at 1mhz vs 4|5mhz - I often get lower total power use at 5mhz than 1mhz (because the active phase finishes more quickly).

While DXCore doesn't officially support it you can switch the clock around yourself. You'll break at least millis(), delay(), and serial while running off an unexpected clock. You might try setting MCLKCTRLB's prescaler bits to the /64 (250khz) or switching the master clock to/from the 32k. Since you are using POWER_DOWN and not STANDBY you are already breaking millis(). You would then switch back on waking up which would make the things you broke work normally again. (You can use delay in the alternate clock you just need to scale your inputs by the ratio of the speeds.)

**Reply by matou78** [reply]
i did a very simple thing to try out the power down mode ,
Just the chip (avr64da32), connected with some little wires to a nrf24l01 module (complete thing with a pcb antenna) and to that i added 3 wires for VCC,GND and updi .

the only thing i didn't tested ( i should have done it first ) is the avr chip alone .
i will try with the different clocks and see what happens with or without NRF24 module .

I use platformIO so it's easy to change clocks , even the unusual ones .

Just i'm not sure how to set the registers myself :( like the MCLKCTRLB for example, can you please explain me with a few words or an example please ?

Thanks for the suggestions and ideas :) will try them asap

**Reply by ObviousInRetrospect** [reply]
try:
```
uint8_t old_div= CLKCTRL.MCLKCTRLB;
 _PROTECTED_WRITE(CLKCTRL.MCLKCTRLB, ((5<<1)|1)); //DIV64 | enable prescale
//... code that runs at 250khz

//restore prescaler
 _PROTECTED_WRITE(CLKCTRL.MCLKCTRLB, old_div);
//back to configured clock
```

for example:
```
 _PROTECTED_WRITE(CLKCTRL.MCLKCTRLB, ((5<<1)|1));
sleep_cpu();
 _PROTECTED_WRITE(CLKCTRL.MCLKCTRLB, old_div);
```

Repeat the power measurements without the nrf - random modules sometimes waste hundreds of ua when they are "off". Or put another sense resistor in line only with the MCU's power pin and/or the nrf module power pin (but then beware leakage on other pins, drive them all low).

When targeting less than a few MA I would assume it might not be the MCU wasting it until I prove otherwise.

Also beware power leakage over the UPDI port, disconnect that for measurement - I've seen that throw readings off by a few hundred UA or worse (depends on the series resistors). For doing power measurements with a serial connection open I use an interposer that uses a pair of Schottky diodes + pull-ups to block current flow over the serial port data lines - BAT20JFILM is a reasonably low leakage option, 0.65ua typ at 5v. 

```
             serial power
                |
mcu tx  --<|--- z------ usb serial rx

                    mcu power
                       |
usb serial tx  --<|--- z------ mcu rx
```
for 115200 I've found 4.7k pull-ups to be fine.
(z is the pull up resistor, -<|- is the diode)

**Reply by matou78** [reply]
thank you a lot ,
i will try it asap ,

for the updi leakage of course, i disconnect it always before measuring

for programmation, i made myself few pcbs with a ch340B and 1k resistor between TX and RX, working nicely

if i can get the avr chip to go as low as i would love to ( i want to try and get the avr chip to work on a CR2532 battery for as long as possible ) , i will surely use a micro mosfet to cut completely the power to the nrf24 module and other tricks to isolate it . but this will be the last thing that i will take care of .

**Comment by SpenceKonde** [comment]
Did you remember to set all the pins to be either output, input_pullup, or input disabled?

The modern AVRs seem to be a lot more sensitive to that than classic ones were.

**Reply by matou78** [reply]
yes , i made all unused pins as pinMode(PIN_XX,INPUT_PULLUP)

**Comment by matou78** [comment]
made few tests with bare chip ( only avr64da32 + 5.6ohm resistor on vcc ) with "loop" empty and all pins set to INPUT_PULLUP
clock i let it 1Mhz and power input used is 3V ( simulating a CR2032 battery)
measured using my trusty multimeter AN8002 on mV setting and of course programmer disconnected

So , normal, not sleeping the resistor measure 5 mV that give me 0,892857mA
SLEEP_MODE_IDLE is the same as normal

SLEEP_MODE_STANDBY at power on, resistor measure 0.6mV that give me 0,107143mA but goes to 0 after around 2-3 sec
SLEEP_MODE_PWR_DOWN at power on ,resistor measure 0.4mV that give me 0,071429mA but goes to 0 after around 2-3 sec

will have to try some code and a button to see how it goes .
then will add the nrf24l01 module and see wtf is eating all that juice .

ObviousInRetrospect , i haven't tried your proposition yet as right now with only the chip and empty , the low power seems to work ,

**Reply by SpenceKonde** [reply]
Might you post the code used to generate this data?

**Reply by matou78** [reply]
Sure , here is the code 
```
#include <Arduino.h>
#include <avr/sleep.h>

void setup() {
  // put your setup code here, to run once:

  pinMode(PIN_PA0,INPUT_PULLUP);
  pinMode(PIN_PA1,INPUT_PULLUP);
  pinMode(PIN_PA2,INPUT_PULLUP);
  pinMode(PIN_PA3,INPUT_PULLUP);
  pinMode(PIN_PA4,INPUT_PULLUP);
  pinMode(PIN_PA5,INPUT_PULLUP);
  pinMode(PIN_PA6,INPUT_PULLUP);
  pinMode(PIN_PA7,INPUT_PULLUP);
  pinMode(PIN_PC0,INPUT_PULLUP);
  pinMode(PIN_PC1,INPUT_PULLUP);
  pinMode(PIN_PC2,INPUT_PULLUP);
  pinMode(PIN_PC3,INPUT_PULLUP);
  pinMode(PIN_PD4,INPUT_PULLUP);
  pinMode(PIN_PD5,INPUT_PULLUP);
  pinMode(PIN_PD6,INPUT_PULLUP);
  pinMode(PIN_PD7,INPUT_PULLUP);
  pinMode(PIN_PF0,INPUT_PULLUP);
  pinMode(PIN_PF1,INPUT_PULLUP);
  pinMode(PIN_PF2,INPUT_PULLUP);
  pinMode(PIN_PF3,INPUT_PULLUP);
  pinMode(PIN_PF4,INPUT_PULLUP);
  pinMode(PIN_PF5,INPUT_PULLUP);
  pinMode(PIN_PF6,INPUT_PULLUP);
  


  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();
  sleep_cpu();
  
}

void loop() {
  // put your main code here, to run repeatedly:
}
```
and here how it's connected 

![image](https://user-images.githubusercontent.com/20770243/183586979-2ac68ad3-532f-4477-a3a6-a37fc6ae9221.png)

**Reply by matou78** [reply]
oh shit , i went too quick and looks like forgot to disable 3 pins ? after correction , the chip goes to 0 after calling standby or power down modes 
  ```
  pinMode(PIN_PA0,INPUT_PULLUP);
  pinMode(PIN_PA1,INPUT_PULLUP);
  pinMode(PIN_PA2,INPUT_PULLUP);
  pinMode(PIN_PA3,INPUT_PULLUP);
  pinMode(PIN_PA4,INPUT_PULLUP);
  pinMode(PIN_PA5,INPUT_PULLUP);
  pinMode(PIN_PA6,INPUT_PULLUP);
  pinMode(PIN_PA7,INPUT_PULLUP);
  pinMode(PIN_PC0,INPUT_PULLUP);
  pinMode(PIN_PC1,INPUT_PULLUP);
  pinMode(PIN_PC2,INPUT_PULLUP);
  pinMode(PIN_PC3,INPUT_PULLUP);
  pinMode(PIN_PD0,INPUT_PULLUP);
  pinMode(PIN_PD1,INPUT_PULLUP);
  pinMode(PIN_PD2,INPUT_PULLUP);
  pinMode(PIN_PD3,INPUT_PULLUP);
  pinMode(PIN_PD4,INPUT_PULLUP);
  pinMode(PIN_PD5,INPUT_PULLUP);
  pinMode(PIN_PD6,INPUT_PULLUP);
  pinMode(PIN_PD7,INPUT_PULLUP);
  pinMode(PIN_PF0,INPUT_PULLUP);
  pinMode(PIN_PF1,INPUT_PULLUP);
  pinMode(PIN_PF2,INPUT_PULLUP);
  pinMode(PIN_PF3,INPUT_PULLUP);
  pinMode(PIN_PF4,INPUT_PULLUP);
  pinMode(PIN_PF5,INPUT_PULLUP);
  pinMode(PIN_PF6,INPUT_PULLUP);
```

**Comment by matou78** [comment, marked as answer]
Hello back :D
so, i got my remote working pretty good with a pretty low power usage
measures given lower are using a shunt resistor 5.6Ohm

chip settings are
- 1Mhz clock
- bod disabled
- all unused pins set to OUTPUT
- ADC disabled ( found in a discussions this thing but can't find where lol `ADC0.CTRLA &= ~ADC_ENABLE_bm;` )

AVR64DA32 + NRF24L01 connected + all buttons used are set as INPUT_PULLUP and in sleep mode SLEEP_MODE_PWR_DOWN
gives me a consumption of 17.85 µA ( 0.10mV)

when i hit a button , i have a spike of 1.9mA (11mV) as the NRF24L01 is sending infos for the receiver

and then until i release the button, the consumption is 0.38mA (2.15mV)

17.85µA on a new CR2032 ( 240mAh ) will give me around 1.5years of work and my remote control is working until the voltage goes under 1.7V

If someone else is looking for an example and ready to use , here is a clean little example if needed
```
#include <Arduino.h> // if you use platformio , delete in arduinoide
#include <avr/sleep.h>

int val_bt = 0;

void blinkLED() {digitalWrite(PIN_PF1, HIGH);} // power up the led

void pins_disabled ()
{
  ADC0.CTRLA &= ~ADC_ENABLE_bm; // disables the analogRead
  pinMode(PIN_PXX,OUTPUT); //XX your unused pins ( 1 per line )
}

void sleep()
{
  set_sleep_mode(SLEEP_MODE_PWR_DOWN); // power down mode
  digitalWrite(PIN_PF1,LOW); //shuts down the LED
  sleep_enable(); //enabling the sleep mode
  sleep_cpu(); // going dark


   //starting position here after wakeup

  sleep_disable(); // disable the sleep mode
}

void interruptButtons()
{
  attachInterrupt(PIN_PD1,blinkLED,LOW); // interrupt on a button for the wakeup with a blink led action at activation (pulling to GND)
}

void setup()
{
  pinMode(PIN_PF1,OUTPUT);
  pinMode(PIN_PD1,INPUT_PULLUP);
  pins_disabled ();
  interruptButtons();

}
void loop() 
{
  sleep();
  val_bt = digitalRead(PIN_PD1);

  if (val_bt == 0)
  {
    //do things if true ( button pressed )
    delay(100);

    while (val_bt == 0)
    {
      //do things while true ( button pressed )
      val_bt = digitalRead(PIN_PD1);
       
    }
    
  }

}

```

Thank you a lot for the time and ideas you gave me :) helped me a lot to think

**Reply by markusschweitzer** [reply]
Hi, how have you connected the NRF24L01? I thought hardware serial is currently broken?!

**Reply by SpenceKonde** [reply]
NRF24 uses SPI? 

And he could always have gotten serial fully working by going back to the last 1.3.x release.

**Reply by matou78** [reply]
hey, yeah the NRF24L01 is using SPI + few control pins, I connected it like on the screenshot below
![image](https://user-images.githubusercontent.com/20770243/188479065-992c63a4-e85b-42d7-afad-17bc26f4e1b5.png)

And I'm using platformIO with a manually updated dxcore to 1.4.7 ( corrected version, my bad)
Never had problems with serial on AVRXXDA chips

**Comment by SpenceKonde** [comment]
How did you get 1.7.4 O_O As the creator of DxCore, I think I'm pretty well
qualified to say that there is no DxCore 1.7.4, the latest version is the
in-development 1.5.0

On Mon, Sep 5, 2022 at 11:15 AM matou78 ***@***.***> wrote:

> hey, yeah the NRF24L01 is using SPI + few control pins, I connected it
> like on the screenshot below
> [image: image]
> <https://user-images.githubusercontent.com/20770243/188479065-992c63a4-e85b-42d7-afad-17bc26f4e1b5.png>
>
> And I'm using platformIO with a manually updated dxcore to 1.7.4
> Never had problems with serial on AVRXXDA chips
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/DxCore/discussions/308#discussioncomment-3561020>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEWYMAM3RUDNHEXZGWKDV4YFAVANCNFSM55AULRCQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by matou78** [reply]
ah fuck me, i'm dumb x) sorry it's 1.4.7 , i updated it from 1.4.5 which was used in the merge of dxcore into platformio
